### PR TITLE
Running more in CI

### PR DESCRIPTION
**Pull request by whytheplatypus**
This PR ...

Adds linting and other checks to the CI checks and moves some checks into parallel jobs.

Test Steps:
1. Check jobs run in this PR e.g. https://github.com/CDCgov/prime-reportstream/pull/2220/checks?check_run_id=3624351115

## Changes
- Adds klint linting check to CI checks
- Moves docs check and linting into parallel jobs (results in longest check running in roughly 11min instead of 16min, and failing much faster)
- Adds Smoke tests

## To Be Done
- [ ] update github settings to require more CI jobs to pass

### CODE REVIEW

File: .github/workflows/build_hub.yml [hunk 1]
```diff
@@ -27,15 +27,15 @@ jobs:
               - 'prime-router/**'
               - '.github/workflows/build_hub.yml'
 
-  build_router:
+  unit_tests:
```
Comment (carlosfelix2) [inline .github/workflows/build_hub.yml:30]: Why not keep the job and name the same?  Why is the next line different?

File: .github/workflows/build_hub.yml [hunk 2]
```diff
@@ -141,5 +155,43 @@ jobs:
...
+    name: Build docker image
+    needs:
+      - unit_tests
+      - linting
+      - docs
+    if: ${{ needs.pre_job.outputs.has_router_change == 'true' }}
+    runs-on: ubuntu-latest
+    env:
+      # These are for CI and not credentials of any system
+      POSTGRES_USER: prime
```
Comment (carlosfelix2) [inline .github/workflows/build_hub.yml:168]: You are duplicating the setting of these values.  I say define a variable early on and reuse it.

File: .github/workflows/build_hub.yml [hunk 3]
```diff
@@ -91,6 +76,35 @@ jobs:
...
+    defaults:
+      run:
+        working-directory: prime-router
+
+    steps:
+      - name: "Check out changes"
+        uses: actions/checkout@v2
+
+      - name: Run Linting Check
+        run: bash ./build.sh -- gradle ktlintCheck
```
Comment (carlosfelix2) [inline .github/workflows/build_hub.yml:94]: I have no such task in my branch.  Where is this coming from?

**Review by carlosfelix2 — CHANGES_REQUESTED** [review]

File: .github/workflows/build_hub.yml (hunk 3 above)
Comment (whytheplatypus) [reply .github/workflows/build_hub.yml:94]: it's setup [here](https://github.com/CDCgov/prime-reportstream/blob/master/prime-router/build.gradle.kts#L33) and the including the plugin [automatically adds the tasks](https://github.com/JLLeitschuh/ktlint-gradle)

**Comment by carlosfelix2** [comment]
Change the title of the PR as this is not an experiment.

**Comment by ronaldheft-usds** [comment]
FYI, @jorg3lopez is always working on running ktlint as a pre-commit hook: https://github.com/CDCgov/prime-reportstream/pull/1965

File: .github/workflows/build_hub.yml [hunk 4]
```diff
@@ -141,5 +155,43 @@ jobs:
         run: |
           false
 
+  build_docker:
+    name: Build docker image
+    needs:
+      - unit_tests
+      - linting
+      - docs
```
Comment (ronaldheft-usds) [inline .github/workflows/build_hub.yml:163]: I'm confused with the addition of all these lines in the job (158-195). These don't seem to have anything to do with linting?

File: .github/workflows/build_hub.yml [hunk 5]
```diff
@@ -27,15 +27,15 @@ jobs:
...
     name: Build Router
     needs: pre_job
     if: ${{ needs.pre_job.outputs.has_router_change == 'true' }}
     runs-on: ubuntu-latest
     env:
       # These are for CI and not credentials of any system
-      POSTGRES_USER: prime
-      POSTGRES_PASSWORD: changeIT!
+      DB_USER: prime
+      DB_PASSWORD: changeIT!
```
Comment (ronaldheft-usds) [inline .github/workflows/build_hub.yml:38]: Why was this changed?

**Review by ronaldheft-usds — CHANGES_REQUESTED** [review]

File: .github/workflows/build_hub.yml (hunk 4 above)
Comment (whytheplatypus) [reply .github/workflows/build_hub.yml:164]: I also split some of the jobs out to allow them to run in parallel. This tells github actions to require these three jobs to pass before starting this one.

btw, is the reason for this job just to check that docker builds?

File: .github/workflows/build_hub.yml (hunk 5 above)
Comment (whytheplatypus) [reply .github/workflows/build_hub.yml]: It allows the removal of the GITHUB_ENV call later.

Comment (whytheplatypus) [reply .github/workflows/build_hub.yml]: eventually we could use github environments to nicely store some of these settings and reduce the conditionals.

File: .github/workflows/build_hub.yml (hunk 4 above)
Comment (ronaldheft-usds) [reply .github/workflows/build_hub.yml]: Ah, I see it now.

Correct, this job verifies the Docker container builds along with all unit tests.

File: .github/workflows/build_hub.yml [hunk 6]
```diff
@@ -141,5 +156,37 @@ jobs:
...
+          then
+              echo "Building for the production environment."
+              echo >> $GITHUB_ENV ACR_REPO=pdhprodcontainerregistry.azurecr.io
+              echo >> $GITHUB_ENV PREFIX=pdhprod
+          else
+              echo "Building for the test environment."
+              echo >> $GITHUB_ENV ACR_REPO=pdhstagingcontainerregistry.azurecr.io
+              echo >> $GITHUB_ENV PREFIX=pdhstaging
+          fi
+
```
Comment (ronaldheft-usds) [inline .github/workflows/build_hub.yml:187]: We can actually drop these lines if you if you update line 192 to be something like `CDCgov/reportstream`. The actual image tag name only matters in the release job, so it's kind of pointless for this extra logic in the CI job.

**Review by ronaldheft-usds — APPROVED** [review]
Looking great, with just one suggested change. After you merge this, can you ping me on Slack? I want to updated our required passing CI jobs to include the new parallelized docs job and the linting job.

File: .github/workflows/build_hub.yml (hunk 6 above)
Comment (whytheplatypus) [reply .github/workflows/build_hub.yml:187]: ok makes total sense, that's why I was wondering if it was just checking the build.

**Comment by whytheplatypus** [comment]
@carlosfelix2 ok i think I got everything! but double check me

**Review by jimduff-usds — APPROVED** [review]

**Review by carlosfelix2 — APPROVED** [review]
